lib/bddsm/it.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module BDDSM
+  class It
+    class Exception < StandardError
+      attr_reader :line_code, :context_title
+
+      def initialize(msg, line_code:, context_title:)
+        @line_code = line_code
+        @context_title = context_title
+        super(msg)
+      end
+    end
+
+    def initialize(describe:, &block)
+      @describe = describe
+      @block = block
+    end
+
+    def run
+      instance_eval(&@block)
+    rescue Matcher::Exception => e
```

Comment:
failed assertions should not be exceptions

Reply:
мне стало интересно как сделан сам rspec. я посмотрел самую первую версию 2005 г

и там я нашел понятие Expectations, оно введено вместо Actual, делает примерно то же самое что наш Actual

урезанная версия: 

module Spec
  module Expectations

    def should(message=nil)
      message ||= "Expectation not met."
      if (! yield)
        raise Spec::Exceptions::ExpectationNotMetError.new(message)
      end
    end


    def should_not_include(sub, message=nil)
      message ||= "<#{self.inspect}> should not include <#{sub.inspect}>"
      should(message) { not self.include? sub }
    end
    
  end

end

Так же райзит ошибку при не выполнении ожидания
У них там нет ни дискрайб, ни ит, есть понятие контекста, которое заменяет эти два понятия


module Spec
  class Context
  
    def initialize(specification=nil)
      @specification = specification
    end

    def run(result_listener)
      result = false
    
      result_listener.spec(@specification)
      setup
      begin
        __send__(@specification)
        result_listener.pass(@specification)
      rescue Exception
        result_listener.failure(@specification, $!)
      end
      teardown
      
      return result
    end
    
    и сюрприз сюрприз, они обертывают ран в обработчик ошибки, которые регистрирует failure.
    
    Это какой-то антипаттерн? Почему так нельзя делать?

Reply:
https://rubygems.org/gems/rspec/versions/0.1.0

Reply:
я именно смотрел пилотную версию, когда там было минимум функционала, чтобы быстро разобраться. Новые версии, к сожалению не смотрел, как у них это работает

Reply:
@sunnmas в этом репо мы отрабатываем общение на английском. Прошу, не игнорируй это.

Reply:
Yes, using exceptions as control flow is antipattern in common, because it makes flow unpredicatable as GOTO. In this case looks like you are right. Exceptions are just fine. We are going to have some issues to collect multiple failed expectations inside single spec. But let's keep it simple.